File: ml/risk_engine.py
from enum import Enum

class RiskLevel(Enum):
    LOW = 1
    MODERATE = 2
    HIGH = 3

class RiskAssessment:
    def __init__(self, level, recommendation, why_explanation, how_explanation):
        self.level = level
        self.recommendation = recommendation
        self.why_explanation = why_explanation
        self.how_explanation = how_explanation

def evaluate_risk(closest_distance, is_closing_in, current_speed, is_raining, is_night):
    # This mirrors the Dart logic we wrote for ContextDrive

    if closest_distance < 0:
        closest_distance = 1.0  # default safe distance

    # Proximity thresholds
    critical_distance = 0.6  # Alerts when vehicle takes up about ~30% of screen
    warning_distance = 0.85  # Alerts when vehicle takes up about ~10% of screen

    # Baseline logic
    if closest_distance <= critical_distance and is_closing_in:
        return RiskAssessment(
            RiskLevel.HIGH,
            "BRAKE NOW",
            f"Object dangerously close ({closest_distance:.2f}) and closing in rapidly.",
            "Visual proximity breached threshold."
        )

    if current_speed > 60:
        if closest_distance <= warning_distance:
            return RiskAssessment(
                RiskLevel.HIGH,
                "Slow down immediately!",
                "High speed combined with close proximity is highly dangerous.",
                "Speed + Proximity rule triggered."
            )
        
        if is_raining:
            return RiskAssessment(
                RiskLevel.MODERATE,
                "Reduce speed due to rain.",
                "Wet roads increase braking distance at high speeds.",
                "Speed + Weather rule triggered."
            )

    if closest_distance <= warning_distance:
        level = RiskLevel.MODERATE
        if is_raining or is_night:
            level = RiskLevel.HIGH
        
        return RiskAssessment(
            level,
            "Increase following distance.",
            "Vehicle ahead is too close.",
            "Proximity warning."
        )

    return RiskAssessment(
        RiskLevel.LOW,
        "Normal Driving.",
        "Conditions are safe.",
        "No risk factors detected."
    )

File: app/lib/engine/risk_engine.dart
import '../models/context_vector.dart';

class RiskEngine {
  /// Deterministic rule engine to calculate risk level based on the current ContextVector
  RiskAssessment assessRisk(ContextVector context) {
    RiskLevel level = RiskLevel.low;
    String how = "Normal driving conditions.";
    String why = "Your speed and environment are stable.";
    String recommendation = "Continue driving safely.";

    // Example deterministic rules based on PRD
    bool hasSpeed = context.currentSpeed != null;
    bool highSpeed = hasSpeed && context.currentSpeed! > 60.0;
    bool veryHighSpeed = hasSpeed && context.currentSpeed! > 80.0;
    bool poorVisibility = context.isNight || (context.isWeatherAvailable && (context.isRaining || context.visibility < 1000));
    
    // Convert generic distance to categories based on our mapping in RiskManager
    bool isVeryNear = context.closestVehicleDistance <= 0.1;
    bool isNear = context.closestVehicleDistance <= 0.3 && !isVeryNear;

    if (isVeryNear && context.isClosingIn) {
      level = RiskLevel.high;
      how = "Risk increased because a vehicle is very close ahead and closing in.";
      why = "You are closing in rapidly and have minimal time to react.";
      recommendation = "Brake immediately to increase following distance.";
    } else if ((isNear && context.isClosingIn) || (isVeryNear)) {
      level = RiskLevel.moderate;
      how = "A vehicle is near and closing in, or very near.";
      why = "You have less time to react if the vehicle in front brakes abruptly.";
      recommendation = "Reduce speed and increase following distance.";
    } else if (veryHighSpeed && poorVisibility) {
      level = RiskLevel.moderate;
      how = "Very high speed in poor visibility conditions.";
      why = "You cannot see far enough ahead to stop safely at this speed.";
      recommendation = "Reduce speed to match visibility conditions.";
    } else if (!hasSpeed) {
      level = RiskLevel.limited;
      how = "Sensor data limited.";
      why = "GPS speed is unavailable or stale.";
      recommendation = "Drive with caution; assistance features limited.";
    }

    return RiskAssessment(
      level: level,
      howExplanation: how,
      whyExplanation: why,
      recommendation: recommendation,
    );
  }
}
